modal/_serialization.py

```diff
@@ -411,75 +412,112 @@ class ParamTypeInfo:
+    api_pb2.PARAM_TYPE_STRING: ParamTypeInfo(
+        default_field="string_default",
+        proto_field="string_value",
+        converter=str,
+        type=str,
+    ),
+    api_pb2.PARAM_TYPE_INT: ParamTypeInfo(
+        default_field="int_default",
+        proto_field="int_value",
+        converter=int,
+        type=int,
+    ),
     api_pb2.PARAM_TYPE_BYTES: ParamTypeInfo(
-        default_field="bytes_default", proto_field="bytes_value", converter=assert_bytes
+        default_field="bytes_default",
+        proto_field="bytes_value",
+        converter=assert_bytes,
+        type=bytes,
     ),
 }
 
 
-def serialize_proto_params(python_params: dict[str, Any], schema: typing.Sequence[api_pb2.ClassParameterSpec]) -> bytes:
-    proto_params: list[api_pb2.ClassParameterValue] = []
+def apply_defaults(
+    python_params: typing.Mapping[str, Any], schema: typing.Sequence[api_pb2.ClassParameterSpec]
+) -> dict[str, Any]:
+    """Apply any declared defaults from the provided schema, if values aren't provided in python_params
+
+    Conceptually similar to inspect.BoundArguments.apply_defaults.
+
+    Note: Apply this before serializing parameters in order to get consistent parameter
+        pools regardless if a value is explicitly provided or not.
+    """
+    result = {**python_params}
     for schema_param in schema:
-        type_info = PROTO_TYPE_INFO.get(schema_param.type)
-        if not type_info:
-            raise ValueError(f"Unsupported parameter type: {schema_param.type}")
+        if schema_param.has_default and schema_param.name not in python_params:
+            default_field_name = schema_param.WhichOneof("default_oneof")
+            if default_field_name is None:
+                raise InvalidError(f"{schema_param.name} declared as having a default, but has no default value")
+            result[schema_param.name] = getattr(schema_param, default_field_name)
+    return result
+
+
+def serialize_proto_params(python_params: dict[str, Any]) -> bytes:
+    proto_params: list[api_pb2.ClassParameterValue] = []
+    for param_name, python_value in python_params.items():
+        python_type = type(python_value)
+        protobuf_type = validate_parameter_type(python_type)
+        type_info = PROTO_TYPE_INFO.get(protobuf_type)
         proto_param = api_pb2.ClassParameterValue(
-            name=schema_param.name,
-            type=schema_param.type,
+            name=param_name,
+            type=protobuf_type,
         )
-        python_value = python_params.get(schema_param.name)
-        if python_value is None:
-            if schema_param.has_default:
-                python_value = getattr(schema_param, type_info.default_field)
-            else:
-                raise ValueError(f"Missing required parameter: {schema_param.name}")
         try:
             converted_value = type_info.converter(python_value)
```

Comment:
I'm a little confused about the role of the `converter`. If we're calling the converter during *serialization*, and the converter is typically itself the type, what type do we expect the input to have here?

Reply:
Yeah I don't think this is very consistent, and I don't particularly like the converter name either. Will redo this code in a followup PR to be more like the type handling in the "polyglot" draft PR where register individual handler classes for each payload type and let those group information about serialization/deserialization logic for each type

Reply:
(Currently we do expect the input types to be of the expected type, and the "converters" are a way to enforce the type if the provided value is *not* of the right type, like converting strings into ints, but I think it would actually be better to crash in that case...)